server/api/router/post.ts

```diff
@@ -275,99 +295,54 @@ export const postRouter = createTRPCRouter({
       const limit = input?.limit ?? 50;
       const { cursor, sort, tag, searchTerm } = input;
 
-      const orderMapping = {
-        newest: {
-          published: "desc" as Prisma.SortOrder,
-        },
-        oldest: {
-          published: "asc" as Prisma.SortOrder,
-        },
-        top: {
-          likes: {
-            _count: "desc" as Prisma.SortOrder,
-          },
-        },
-      };
-      const orderBy = orderMapping[sort] || orderMapping["newest"];
-
-      const response = await ctx.prisma.post.findMany({
-        take: limit + 1,
-        where: {
-          published: {
-            lte: new Date(),
-            not: null,
-          },
-          ...(tag
-            ? {
-                tags: {
-                  some: {
-                    tag: {
-                      title: {
-                        contains: tag?.toUpperCase() || "",
-                      },
-                    },
-                  },
-                },
-              }
-            : {}),
-          ...(searchTerm
-            ? {
-                OR: [
-                  {
-                    user: {
-                      name: {
-                        contains: searchTerm || "",
-                        mode: "insensitive",
-                      },
-                    },
-                  },
-                  {
-                    title: {
-                      contains: searchTerm || "",
-                      mode: "insensitive",
-                    },
-                  },
-                  {
-                    excerpt: {
-                      contains: searchTerm || "",
-                      mode: "insensitive",
-                    },
-                  },
-                ],
-              }
-            : {}),
-        },
-        select: {
+      const posts = await ctx.db.query.post.findMany({
+        columns: {
           id: true,
           title: true,
-          updatedAt: true,
           published: true,
           readTimeMins: true,
           slug: true,
           excerpt: true,
+        },
+        with: {
+          bookmarks: true,
+          likes: true,
           user: {
-            select: { name: true, image: true, username: true },
-          },
-          bookmarks: {
-            select: { userId: true },
-            where: { userId: userId },
+            columns: {
+              name: true,
+              image: true,
+              username: true,
+            },
           },
         },
-        cursor: cursor ? { id: cursor } : undefined,
-        skip: cursor ? 1 : 0,
-        orderBy,
+        where: and(
+          tag
+            ? (lte(post.published, new Date()),
+              eq(tagSchema.title, tag.toUpperCase()))
+            : lte(post.published, new Date()),
```

Comment:
Heres why your cursor isnt working I think. 

Depending on the sort option selected it should either by

```sort === "newest" lte(post.published, cursor.published ?? new Date()```
```sort === "oldest" gte(post.published, cursor.published ?? new Date(0)```

... sort === "top" I am not sure because you dont have a sequential column.. 

```sort === "top" lte(count(likeSchema.id), cursor.likes ?? MAX_INT ```

Is my initial thought but havent tested

Reply:
Not sure I understand the suggestion tbh. Will look again tomorrow with fresh eyes because it's beat me today.